### PR TITLE
(PUP-3319) Retry to send report if connection to master fails

### PR DESCRIPTION
**Pull request by roidelapluie**

### CODE REVIEW

**Comment by joshcooper** [comment]
The retry mechanism has been added back in https://github.com/puppetlabs/puppet/commit/b8370b112ee96f4aedec59ab7303b786a96a8af2.

However, Puppet uses a PUT request to "save" the report. PUT requests are supposed to be [idempotent](https://tools.ietf.org/html/rfc7231#page-23):

> A request method is considered "idempotent" if the intended effect on the server of multiple identical requests with that method is the same as the effect for a single such request.

But each time we PUT a report, the server stores a new copy on disk and in puppetdb. In order to idempotent, we would need to update the *one* report resource.

So if we were to retry sending of reports, I think we would need to be very restrictive about the kinds of exceptions we were to retry on. For example, I think `Errno::ECONNREFUSED` would be ok, since we know that we couldn't connect to the report server.

It appears that ruby 2.1.5 will automatically [retry idempotent http requests](https://github.com/ruby/ruby/blob/v2_1_5/lib/net/http.rb#L1421-L1431). It is concerning that it retries ReadTimeout, because the server could be taking a long time to process the report and the agent times out waiting for the response, causing the agent to resend the report.

File: lib/puppet/configurer.rb [hunk 1]
```diff
@@ -3,6 +3,7 @@
 require 'timeout'
 require 'puppet/network/http_pool'
 require 'puppet/util'
+require 'puppet/util/retryaction'
```
Comment (MikaelSmith) [inline lib/puppet/configurer.rb]: The PR's been merged, with the file name changed to retry_action as Rob had suggested.

**Comment by MikaelSmith** [comment]
Triage Notes:
* This isn't ready to merge. It needs a rebase, and Josh's comments need to be addressed about retrying non-idempotent actions.
* @MikaelSmith, @joshcooper, and @kylog to review impact of retry on the PUT action, particularly the impact of Ruby 2.1.5.

**Comment by joshcooper** [comment]
Filed https://tickets.puppetlabs.com/browse/PUP-3905 to investigate ruby's automatic retry behavior.

**Comment by kylog** [comment]
This needs a rebase now, possibly due to the change to use rspec 3.

**Comment by joshcooper** [comment]
PR Triage Notes:

* @roidelapluie thank you for your contribution!
* We have confirmed that ruby 2.1.5 will unconditionally retry idempotent methods once due to wide variety of exception types.
* For next steps, we need to decide whether we should retry report saves?
* And if so, what exception types we should rescue. For example, `Errno::ECONNREFUSED` seems like a good candidate, but `ReadTimeout` is not.
* @dankreek is going to take a look and make a proposal for next steps.

**Comment by roidelapluie** [comment]
On Wed, Feb 04, 2015 at 11:05:34AM -0800, Josh Cooper wrote:
> PR Triage Notes:
> 
> * @roidelapluie thank you for your contribution!
> * We have confirmed that ruby 2.1.5 will unconditionally retry idempotent methods once due to wide variety of exception types.
> * For next steps, we need to decide whether we should retry report saves?
> * And if so, what exception types we should rescue. For example, `Errno::ECONNREFUSED` seems like a good candidate, but `ReadTimeout` is not.
> * @dankreek is going to take a look and make a proposal for next steps.
> 

Thank you guys. I'm exited to see it supported in puppet. The ball is in
your side now.


-- 
 (o-    Julien Pivotto
 //\    Open-Source Consultant
 V_/_   Inuits - https://www.inuits.eu

**Comment by kylog** [comment]
https://tickets.puppetlabs.com/browse/PUP-3905 is the ticket to look into 2.1.5 behavior.

I think we should block this on resolution of PUP-3905.

**Comment by kylog** [comment]
Marking as blocked since we want to resolve PUP-3905 first.

**Comment by roidelapluie** [comment]
Not needed anymore

**Comment by kylog** [comment]
@roidelapluie why is this not needed any more? 

No one has made progress on PUP-3905 AFAIK, but I was just wondering if you learned something that obviated the need for this.

**Comment by roidelapluie** [comment]
Now the service that restarts the puppetserver actually waits for the service to be actually restarted. So it it is not needed for my specific use case.

**Pull request closed without merging**